File: problems/0002_add-two-numbers/Solution.scala
/**
 * Definition for singly-linked list.
 * class ListNode(var _x: Int = 0) {
 *   var next: ListNode = null
 *   var x: Int = _x
 * }
 */
object Solution {
    def addTwoNumbers(l1: ListNode, l2: ListNode): ListNode = {
        var (_l1, _l2) = (l1, l2)
        val head = ListNode(0)
        var ptr = head
        var carry = 0
        while (_l1 != null || _l2 != null) {
            val sum = (if (_l1 != null) {
                val x = _l1.x
                _l1 = _l1.next
                x
            } else 0) + (if (_l2 != null) {
                val x = _l2.x
                _l2 = _l2.next
                x
            } else 0) + carry
            ptr.next = ListNode(sum % 10)
            ptr = ptr.next
            carry = (if (sum >= 10) 1 else 0)
        }
        if (carry == 1)
            ptr.next = ListNode(1)
        head.next
    }
}

File: problems/0002_add-two-numbers/Solution.cs
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public int val;
 *     public ListNode next;
 *     public ListNode(int x) { val = x; }
 * }
 */
public class Solution {
    public ListNode AddTwoNumbers(ListNode l1, ListNode l2) {
        ListNode head = new ListNode(0),
                ptr = head;
        int carry = 0;
        for (bool g1 = (l1 != null), g2 = (l2 != null);
                g1 || g2;
                g1 = (l1 != null), g2 = (l2 != null)) {
            int sum = (g1 ? l1.val : 0) + (g2 ? l2.val : 0) + carry;
            ptr.next = new ListNode(sum % 10);
            ptr = ptr.next;
            carry = (sum >= 10 ? 1 : 0);
            if (g1)
                l1 = l1.next;
            if (g2)
                l2 = l2.next;
        }
        if (carry == 1)
            ptr.next = new ListNode(1);
        return head.next;
    }
}
